extensions/ql-vscode/src/vscode-tests/test-config.ts

```diff
@@ -0,0 +1,93 @@
+import { ConfigurationTarget, workspace } from 'vscode';
+import { CUSTOM_CODEQL_PATH_SETTING, REMOTE_CONTROLLER_REPO, REMOTE_REPO_LISTS, Setting } from '../config';
+
+class TestSetting<T> {
+  private settingState: {
+    globalValue?: T;
+    workspaceValue?: T,
+    workspaceFolderValue?: T,
+  } | undefined;
+
+  constructor(
+    public readonly setting: Setting,
+    private initialTestValue: T | undefined = undefined
+  ) { }
+
+  public async get(): Promise<T | undefined> {
+    return this.section.get(this.setting.name);
+  }
+
+  public async set(value: T | undefined, target: ConfigurationTarget = ConfigurationTarget.Global): Promise<void> {
+    await this.section.update(this.setting.name, value, target);
+  }
+
+  public async setInitialTestValue(value: T | undefined) {
+    this.initialTestValue = value;
+  }
+
+  public async initialSetup() {
+    this.settingState = this.section.inspect(this.setting.name);
+
+    // Unfortunately it's not well-documented how to check whether we can write to a workspace
+    // configuration. This is the best I could come up with. It only fails for initial test values
+    // which are not undefined.
+    if (this.settingState?.workspaceValue !== undefined) {
+      await this.set(this.initialTestValue, ConfigurationTarget.Workspace);
+    }
+    if (this.settingState?.workspaceFolderValue !== undefined) {
+      await this.set(this.initialTestValue, ConfigurationTarget.WorkspaceFolder);
+    }
+
+    await this.setup();
+  }
+
+  public async setup() {
+    await this.set(this.initialTestValue, ConfigurationTarget.Global);
+  }
+
+  public async restoreToInitialValues() {
+    const state = this.section.inspect(this.setting.name);
+
+    // We need to check the state of the setting before we restore it. This is less important for the global
+    // configuration target, but the workspace/workspace folder configuration might not even exist. If they
+    // don't exist, VSCode will error when trying to write the new value (even if that value is undefined).
+    if (state?.globalValue !== this.settingState?.globalValue) {
+      await this.set(this.settingState?.globalValue, ConfigurationTarget.Global);
+    }
+    if (state?.workspaceValue !== this.settingState?.workspaceValue) {
```

Comment:
Should this and the check for `workspaceFolderValue` use the same condition as in `initialsetup`? I think either should be fine as they'll have the same effect and both protect against the workspace configuration not existing, having the code be identical could be a nice safety feature.
```suggestion
    if (this.settingState?.workspaceValue !== undefined) {
```

Reply:
I agree that it would be really useful to have identical conditions. However, I don't think this results in identical behavior. For example, if a test is run in a workspace without a `codeQL.canary` setting and a test adds this setting to the workspace, the current behavior would be to remove the `codeQL.canary` flag at the end of the test (since the `state?.workspaceValue === true` and `this.settingState?.workspaceValue === undefined`). With the suggested change, it would not reset the value (since `this.settingState?.workspaceValue === undefined`). So, I don't think this change is safe to make.

Reply:
Ah I see what you mean. Thanks for explaining. Then I agree let's keep it as it is.